daemons/pacemakerd/pcmkd_messages.c

```diff
@@ -39,44 +57,52 @@ pcmk_handle_ping_request(pcmk__client_t *c, xmlNode *msg, uint32_t id)
                    (long long) subdaemon_check_progress);
     crm_xml_add(ping, XML_PING_ATTR_STATUS, "ok");
     reply = create_reply(msg, ping);
+
     free_xml(ping);
-    if (reply) {
-        if (pcmk__ipc_send_xml(c, id, reply, crm_ipc_server_event) !=
-                pcmk_rc_ok) {
-            crm_err("Failed sending ping reply to client %s",
-                    pcmk__client_name(c));
-        }
-        free_xml(reply);
+
+    pcmk__ipc_send_ack(request->ipc_client, request->ipc_id, request->ipc_flags,
+                       "ack", NULL, CRM_EX_INDETERMINATE);
+
+    if (reply == NULL) {
+        pcmk__format_result(&request->result, CRM_EX_ERROR, PCMK_EXEC_ERROR,
+                            "Failed building ping reply for client %s",
+                            pcmk__client_name(request->ipc_client));
     } else {
-        crm_err("Failed building ping reply for client %s",
-                pcmk__client_name(c));
+        pcmk__set_result(&request->result, CRM_EX_OK, PCMK_EXEC_DONE, NULL);
     }
+
     /* just proceed state on sbd pinging us */
     if (from && strstr(from, "sbd")) {
         if (pcmk__str_eq(pacemakerd_state, XML_PING_ATTR_PACEMAKERDSTATE_SHUTDOWNCOMPLETE, pcmk__str_none)) {
             if (pcmk__get_sbd_sync_resource_startup()) {
                 crm_notice("Shutdown-complete-state passed to SBD.");
             }
-            shutdown_complete_state_reported_to = c->pid;
+
+            shutdown_complete_state_reported_to = request->ipc_client->pid;
```

Comment:
Not a problem, just noticing: Before, this global wouldn't be set until after the reply was sent, but with the new model, the reply will be sent after the handler returns. It doesn't matter in this case because even previously, this would be set regardless of whether the reply succeeded or not. But it's an interesting case that we may have to think about at some point (a post-reply handler?).

Reply:
I was a little concerned about this area.  I don't like setting the mainloop trigger before sending the reply.  I worry that this introduces some race condition where the trigger fires before the reply gets a chance to happen.  So yeah, adding a post-reply handler might make me feel a lot better.

Reply:
Setting the startup trigger early is fine, because it's just saying we received the ping from sbd, not that we sent anything to sbd. It's sbd giving us the all-clear to start everything, which is good to go regardless of whether we can ack it.